File: backend/app/github_client.py
import hashlib
import requests
import time

from app.cache import get_json, set_json
from app.config import GITHUB_TOKEN


HEADERS = {"Authorization": f"token {GITHUB_TOKEN}"} if GITHUB_TOKEN else {}

def _timed_get(url, *, headers=None, timeout=None, label="request"):
    start_time = time.time()
    try:
        response = requests.get(url, headers=headers, timeout=timeout)
        elapsed = time.time() - start_time
        print(
            f"[GITHUB_API] {label} -> status={response.status_code} "
            f"time={elapsed:.2f}s"
        )
        return response
    except requests.Timeout:
        elapsed = time.time() - start_time
        print(f"[GITHUB_API] {label} -> timeout after {elapsed:.2f}s")
        raise
    except Exception as exc:
        elapsed = time.time() - start_time
        print(f"[GITHUB_API] {label} -> failed after {elapsed:.2f}s: {exc}")
        raise

def search_repositories(query, per_page=50, max_retries=3):
    """
    Search GitHub repositories with rate limit handling.
    per_page: Number of results per page (max 100, default 30)
    max_retries: Maximum number of retries on rate limit
    """
    print(f"[GITHUB_API] search query={query}")
    cache_key = f"v1:github-search:{hashlib.sha256(f'{query}:{per_page}'.encode()).hexdigest()}"
    cached = get_json(cache_key)
    if cached is not None:
        return cached

    url = f"https://api.github.com/search/repositories?q={query}&sort=stars&per_page={per_page}"
    
    for attempt in range(max_retries):
        response = _timed_get(
            url,
            headers=HEADERS,
            label=f"search_repositories attempt={attempt + 1}/{max_retries}"
        )
        
        if response.status_code == 200:
            data = response.json()
            items = data.get("items", [])
            total_count = data.get("total_count", 0)
            remaining = response.headers.get("X-RateLimit-Remaining", "unknown")
            limit = response.headers.get("X-RateLimit-Limit", "unknown")
            print(f"  -> Found {len(items)} repos (total: {total_count}, rate limit: {remaining}/{limit})")
            set_json(cache_key, items, ttl=1200)
            return items
        elif response.status_code == 403:
            # Rate limit exceeded
            error_data = response.json() if response.text else {}
            error_msg = error_data.get("message", "Rate limit exceeded")
            
            # Check if there's a retry-after header
            retry_after = response.headers.get("Retry-After")
            if retry_after:
                wait_time = int(retry_after) + 1
                print(f"  -> Rate limit exceeded. Waiting {wait_time}s before retry {attempt + 1}/{max_retries}...")
                time.sleep(wait_time)
                continue
            else:
                # Default wait time if no retry-after header
                wait_time = (attempt + 1) * 10  # Exponential backoff
                print(f"  -> Rate limit exceeded. Waiting {wait_time}s before retry {attempt + 1}/{max_retries}...")
                time.sleep(wait_time)
                continue
        else:
            print(f"  -> Error: status={response.status_code}")
            return []
    
    # All retries exhausted
    print(f"  -> Rate limit exceeded after {max_retries} retries. Returning empty results.")
    return []

def get_all_open_issues(owner, repo, timeout=5):
    """Fetch all open issues without label filter. Added timeout to prevent hanging."""
    cache_key = f"v1:issues:{owner.lower()}/{repo.lower()}"
    cached = get_json(cache_key)
    if cached is not None:
        return cached

    url = f"https://api.github.com/repos/{owner}/{repo}/issues?state=open&per_page=20&sort=updated&direction=desc"
    try:
        response = _timed_get(
            url,
            headers=HEADERS,
            timeout=timeout,
            label=f"get_all_open_issues repo={owner}/{repo}"
        )
        if response.status_code == 200:
            issues = response.json()
            # Filter out pull requests (GitHub API returns both issues and PRs)
            # PRs have a 'pull_request' key, issues don't
            issues_only = [issue for issue in issues if 'pull_request' not in issue]
            result = issues_only if isinstance(issues_only, list) else []
            set_json(cache_key, result, ttl=600)
            return result
        return []
    except requests.Timeout:
        return []
    except requests.RequestException:
        return []

File: backend/src/github-client.ts
import { createHash } from "node:crypto";
import { Agent, request } from "undici";
import { getJson, setJson } from "./cache.js";
import { config } from "./config.js";
import type { GitHubIssue, GitHubRepo } from "./types.js";

const agent = new Agent({ connections: 20, connect: { timeout: 2_000 } });
const headers: Record<string, string> = {
  Accept: "application/vnd.github+json",
  "User-Agent": "Repo-Scout",
  ...(config.githubToken ? { Authorization: `Bearer ${config.githubToken}` } : {}),
};

const sleep = (milliseconds: number) => new Promise((resolve) => setTimeout(resolve, milliseconds));
const digest = (value: string) => createHash("sha256").update(value).digest("hex");

async function githubGet<T>(url: string, timeout: number): Promise<{ status: number; data: T; responseHeaders: Record<string, string | string[] | undefined> }> {
  const response = await request(url, {
    method: "GET",
    headers,
    dispatcher: agent,
    headersTimeout: timeout,
    bodyTimeout: timeout,
  });
  const data = await response.body.json() as T;
  return { status: response.statusCode, data, responseHeaders: response.headers };
}

export async function searchRepositories(query: string, perPage = 50, maxRetries = 3): Promise<GitHubRepo[]> {
  console.log(`[GITHUB_API] search query=${query}`);
  const cacheKey = `v1:github-search:${digest(`${query}:${perPage}`)}`;
  const cached = await getJson<GitHubRepo[]>(cacheKey);
  if (cached) {
    console.log(`[GITHUB_API] repository_search source=redis status=hit results=${cached.length}`);
    return cached;
  }
  console.log("[GITHUB_API] repository_search source=github status=cache_miss");

  const url = `https://api.github.com/search/repositories?q=${encodeURIComponent(query)}&sort=stars&per_page=${perPage}`;
  for (let attempt = 1; attempt <= maxRetries; attempt += 1) {
    const started = performance.now();
    try {
      const response = await githubGet<{ items?: GitHubRepo[]; total_count?: number; message?: string }>(url, 8_000);
      console.log(`[GITHUB_API] search_repositories attempt=${attempt}/${maxRetries} -> status=${response.status} time=${secondsSince(started)}s`);
      if (response.status === 200) {
        const items = response.data.items ?? [];
        console.log(`  -> Found ${items.length} repos (total: ${response.data.total_count ?? 0}, rate limit: ${response.responseHeaders["x-ratelimit-remaining"] ?? "unknown"}/${response.responseHeaders["x-ratelimit-limit"] ?? "unknown"})`);
        await setJson(cacheKey, items, 1_200);
        return items;
      }
      if (response.status !== 403) return [];
      const retryHeader = response.responseHeaders["retry-after"];
      const retryAfter = Number(Array.isArray(retryHeader) ? retryHeader[0] : retryHeader);
      await sleep(Number.isFinite(retryAfter) ? (retryAfter + 1) * 1_000 : attempt * 10_000);
    } catch (error) {
      console.log(`[GITHUB_API] search_repositories attempt=${attempt}/${maxRetries} -> failed after ${secondsSince(started)}s: ${error instanceof Error ? error.message : String(error)}`);
      if (attempt === maxRetries) return [];
    }
  }
  return [];
}

export async function getOpenIssues(owner: string, repo: string): Promise<GitHubIssue[]> {
  const cacheKey = `v1:issues:${owner.toLowerCase()}/${repo.toLowerCase()}`;
  const cached = await getJson<GitHubIssue[]>(cacheKey);
  if (cached) return cached;
  try {
    const url = `https://api.github.com/repos/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}/issues?state=open&per_page=20&sort=updated&direction=desc`;
    const response = await githubGet<GitHubIssue[]>(url, 5_000);
    if (response.status !== 200 || !Array.isArray(response.data)) return [];
    const issues = response.data.filter((issue) => !("pull_request" in issue));
    await setJson(cacheKey, issues, 600);
    return issues;
  } catch {
    return [];
  }
}

function secondsSince(started: number): string {
  return ((performance.now() - started) / 1_000).toFixed(2);
}
